rotkehlchen/exchanges/bitfinex.py

```diff
@@ -0,0 +1,933 @@
+            success=was_successful,
+            response=response,
+            currency_map=currency_map,
+        )
+
+    def _query_exchange_pairs(self) -> ExchangePairsResponse:
+        """Query and return the list of the exchange (trades) pairs in
+        `<ExchangePairsResponse>.pairs`.
+        Otherwise populate <ExchangePairsResponse> with data that each endpoint
+        can process as an unsuccessful request.
+        """
+        was_successful = True
+        pairs = []
+        response = self._api_query('configs_list_pair_exchange')
+
+        if response.status_code != HTTPStatus.OK:
+            was_successful = False
+            log.error(f'{self.name} exchange pairs list query failed. Check further logs')
+        else:
+            try:
+                response_list = rlk_jsonloads_list(response.text)
+            except JSONDecodeError:
+                was_successful = False
+                log.error(
+                    f'{self.name} exchange pairs list returned invalid JSON response. '
+                    f'Check further logs',
+                )
+            else:
+                pairs = response_list[0]
+
+        return ExchangePairsResponse(
+            success=was_successful,
+            response=response,
+            pairs=pairs,
+        )
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['validate_api_key'],
+    ) -> Tuple[bool, str]:
+        ...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['balances'],
+    ) -> Tuple[Optional[Dict[Asset, Balance]], str]:
+        ...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['trades'],
+    ) -> List[Trade]:
+        ...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['asset_movements'],
+    ) -> List[AssetMovement]:
+        ...
+
+    def _process_unsuccessful_response(
+            self,
+            response: Response,
+            case: Literal['validate_api_key', 'balances', 'trades', 'asset_movements'],
+    ) -> Union[
+        List,
+        Tuple[bool, str],
+        Tuple[Optional[Dict[Asset, Balance]], str],
+    ]:
+        """This function processes not successful responses for the following
+        cases listed in `case`.
+        """
+        try:
+            response_list = rlk_jsonloads_list(response.text)
+        except JSONDecodeError as e:
+            msg = f'{self.name} returned invalid JSON response: {response.text}.'
+            log.error(msg)
+
+            if case in {'validate_api_key', 'balances'}:
```

Comment:
I think it's better to put these as tuples and not sets (here and elsewhere in the PR). It's probably miniscule but instantiating a tuple is faster than instantiating a set.

Reply:
I'd say true :D? I know search is faster in a set, but as I'm not sure about the difference on instantiation times I will change it.